pilot/pkg/security/authz/builder/builder.go

```diff
@@ -112,61 +153,122 @@ func build(policies []model.AuthorizationPolicy, tdBundle trustdomain.Bundle,
+				authzLog.Errorf("Skipped nil rule %s", name)
 				continue
 			}
-			m, err := authzmodel.New(rule, option.IsIstioVersionGE15)
+			m, err := authzmodel.New(rule, b.option.IsIstioVersionGE15)
 			if err != nil {
-				authzLog.Errorf("skipped rule %s: %v", name, err)
+				authzLog.Errorf("Skipped rule %s: %v", name, err)
 				continue
 			}
-			m.MigrateTrustDomain(tdBundle)
+			m.MigrateTrustDomain(b.trustDomainBundle)
 			generated, err := m.Generate(forTCP, action)
 			if err != nil {
-
-				if forTCP && option.IsOnInboundPassthrough {
-					authzLog.Debugf("On TCP Inbound Passthroush filter chain, skipped rule %s: %v", name, err)
+				if forTCP && b.option.IsOnInboundPassthrough {
+					authzLog.Debugf("On TCP inbound passthrough filter chain, skipped rule %s: %v", name, err)
 				} else {
-					authzLog.Errorf("skipped rule %s: %v", name, err)
+					authzLog.Errorf("Skipped rule %s: %v", name, err)
 				}
-
 				continue
 			}
 			if generated != nil {
 				rules.Policies[name] = generated
-				authzLog.Debugf("rule %s generated policy: %+v", name, generated)
+				authzLog.Debugf("Rule %s generated policy: %+v", name, generated)
 			}
 		}
 	}
 
-	return &rbachttppb.RBAC{Rules: rules}
+	if forTCP {
+		return &builtConfigs{tcp: b.buildTCP(rules, providers)}
+	}
+	return &builtConfigs{http: b.buildHTTP(rules, providers)}
 }
 
-// nolint: interfacer
-func createHTTPFilter(config *rbachttppb.RBAC) *httppb.HttpFilter {
-	if config == nil {
-		return nil
+func (b Builder) buildHTTP(rules *rbacpb.RBAC, providers []string) []*httppb.HttpFilter {
+	if !b.option.IsCustomBuilder {
+		rbac := &rbachttppb.RBAC{Rules: rules}
+		return []*httppb.HttpFilter{
+			{
+				Name:       authzmodel.RBACHTTPFilterName,
+				ConfigType: &httppb.HttpFilter_TypedConfig{TypedConfig: util.MessageToAny(rbac)},
+			},
+		}
 	}
-	return &httppb.HttpFilter{
-		Name:       authzmodel.RBACHTTPFilterName,
-		ConfigType: &httppb.HttpFilter_TypedConfig{TypedConfig: util.MessageToAny(config)},
+
+	extauthz, err := buildExtAuthz(b.extensions, providers)
+	if err != nil {
+		authzLog.Errorf("Failed parsing CUSTOM action, will generate a deny all config: %v", err)
+		rbac := &rbachttppb.RBAC{Rules: rbacDefaultDenyAll}
+		return []*httppb.HttpFilter{
+			{
+				Name:       authzmodel.RBACHTTPFilterName,
+				ConfigType: &httppb.HttpFilter_TypedConfig{TypedConfig: util.MessageToAny(rbac)},
+			},
+		}
+	}
+	// Add the RBAC filter in shadow mode so that it only evaluates the matching rules for CUSTOM action but not enforce it.
+	// The evaluation result is stored in the dynamic metadata keyed by the policy name.
+	// See https://docs.google.com/document/d/1V4mCQCw7mlGp0zSQQXYoBdbKMDnkPOjeyUb85U07iSI/edit#bookmark=kix.jdq8u0an2r6s
+	// for more details.
+	rbac := &rbachttppb.RBAC{ShadowRules: rules}
+	return []*httppb.HttpFilter{
+		{
+			Name:       authzmodel.RBACHTTPFilterName,
+			ConfigType: &httppb.HttpFilter_TypedConfig{TypedConfig: util.MessageToAny(rbac)},
```

Comment:
Does this mean this filter will be executed twice?

Reply:
It depends on your authz policy. If you have Authz policy with both CUSTOM and non-CUSTOM action, then yes. The 1st RBAC will only evaluate the matching rule. We already do this for DENY policy (it will be translated to a different RBAC filter).